compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -287,6 +287,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expr_ty: Ty<'tcx>,
     ) {
         if let ty::Adt(expected_adt, substs) = expected.kind() {
+            if let hir::ExprKind::Field(base, ident) = expr.kind
+                && let Some(typeck_results) = self.in_progress_typeck_results
+            {
+                let base_ty = typeck_results.borrow().expr_ty(base);
+                if base_ty == expected {
+                    err.span_suggestion_verbose(
+                        expr.span.with_lo(base.span.hi()),
```

Comment:
What does this do if `base` comes from a macro expansion? For example:

```rust
// run-rustfix

pub struct MyWrapper(u32);

macro_rules! my_wrapper {
  ($expr:expr) => { MyWrapper($expr) }
}

fn main() {
    some_fn(my_wrapper!(123).0); //~ ERROR mismatched types
}

fn some_fn(wrapped: MyWrapper) {
    drop(wrapped);
}
```

I think `base.span` should probably be adjusted to account for macros. Try matching `if let Some(base_span) 
 = base.span.find_ancestor_inside(expr.span)` first to make sure that expr span and base span overlap.